index.src.html

```diff
@@ -819,93 +819,94 @@ <h3 id=interface-reporting-observer>Interface {{ReportingObserver}}</h3>
 
   2. Remove the <a>context object</a> from the <a>registered reporting observer
      list</a> of its associated <a>environment settings object</a>.
 
   <h3 id="add-report" algorithm>
-    Add |report| to reporting |observer|
+    Add |report| to |observer|
   </h3>
 
-  This algorithm adds |report| as a {{Report}} object to |observer|'s <a>report
-  list</a>, so long as |report|'s <a>type</a> is observable by |observer|.
+  Given a <a>report</a> |report| and a {{ReportingObserver}} |observer|, this
+  algorithm adds |report| to |observer|'s <a>report list</a>, so long as
+  |report|'s <a>type</a> is observable by |observer|.
 
   1. If |observer|'s <a>options</a> has a {{ReportingObserverOptions/types}}
-     member which does not contain |report|'s [=report/type=], stop.
+     member which does not contain |report|'s [=report/type=], return.
 
   2. Create a new {{Report}} |r| with {{Report/type}} initialized to |report|'s
      [=report/type=], {{Report/url}} initialized to |report|'s [=report/url=],
      and {{Report/body}} initialized to |report|'s [=report/body=].
 
   Issue: how to polymorphically initialize body?
 
-  3. Add |r| to the end of |observer|'s <a>report list</a>.
-      
+  3. Append |r| to |observer|'s <a>report list</a>.
+
+  4. If the size of |observer|'s <a>report list</a> is 1, <a>Queue a task</a> to
+     [[#invoke-observers]] with a copy of the <a>registered reporting observer
+     list</a> of the <a>environment settings object</a> associated with
+     |observer|.
+
   <h3 id="notify-observers" algorithm>
     Notify reporting observers on |settings| with |report|
   </h3>
-  
+
   This algorithm makes |report|'s contents available to any <a>registered</a>
   <a>reporting observers</a> on the provided <a>environment settings object</a>
   |settings|.
-  
+
   1. For each {{ReportingObserver}} |observer| <a>registered</a> with
      |settings|, execute [[#add-report]] on |report| and |observer|.
- 
+
   2. Add |report| to the <a>report buffer</a> associated with |settings|.
-  
+
   3. If the <a>report buffer</a> now contains more than 100 reports, remove the
-     oldest report.
-  
-  4. <a>Queue a task</a> to [[#invoke-observers]] with a copy of the
-     <a>registered reporting observer list</a> associated with |settings|.
-   
+     item at the beginning of the <a>report buffer</a>.
+
   <h3 id="invoke-observers" algorthm>
     Invoke reporting observers with |notify list|
   </h3>
-  
+
   This algorithm invokes observer callback functions for reports of previously
   observed behavior.
-  
-  Note: It is possible for the <a>registered</a> reporting observers and
-  their <a>options</a> to be different at this time than when reports were
-  generated. It is important that we rely only on the state at the time of
-  generation, for example to permit libraries to observe behavior only within
-  the context of their own code.
-  
+
   1. For each {{ReportingObserver}} |observer| in |notify list|:
```

Comment:
I think we want to add something here that checks to see if the report list is empty, or else there are still conditions where `callback` will be called with an empty list (Because 5.2 -- which is called every time you `observe()` with a new not-future-only observer -- will queue a task to call this algorithm with a list of all registered observers, the second time this is called, some of the report lists will have been cleared)

Reply:
Okay, I've added a line to continue on empty list.